Sources/App/Core/Search.swift

```diff
@@ -75,57 +82,77 @@ enum Search {
+    static func packageMatchQueryBuilder(on database: Database,
+                                         terms: [String],
+                                         offset: Int? = nil,
+                                         limit: Int? = nil) -> SQLSelectBuilder {
         guard let db = database as? SQLDatabase else {
             fatalError("Database must be an SQLDatabase ('as? SQLDatabase' must succeed)")
         }
 
         let maxSearchTerms = 20  // just to impose some sort of limit
 
         // binds
-        let sanitizedTerms = sanitize(terms)
-        guard !sanitizedTerms.isEmpty else {
-            return nil
-        }
-        let mergedTerms = SQLBind(
-            sanitizedTerms.joined(separator: " ").lowercased()
-        )
-        let binds = sanitizedTerms[
-            ..<min(sanitizedTerms.count, maxSearchTerms)
-        ].map(SQLBind.init)
+        let binds = terms[..<min(terms.count, maxSearchTerms)].map(SQLBind.init)
+        let mergedTerms = SQLBind(terms.joined(separator: " ").lowercased())
 
         // constants
         let empty = SQLLiteral.string("")
-        let space = SQLLiteral.string(" ")
         let contains = SQLRaw("~*")
 
         let haystack = concat(
-            packageName, space, coalesce(summary, empty), space, repoName, space, repoOwner
+            with: " ",
+            packageName, coalesce(summary, empty), repoName, repoOwner
         )
+        let sortOrder = SQLOrderBy(eq(lower(packageName), mergedTerms),
+                                   .descending)
+            .then(score, .descending)
+            .then(packageName, .ascending)
 
         let preamble = db
             .select()
-            .column(packageMatch)
-            .column(id)
+            .column(.package)
+            .column(null, as: keyword)
+            .column(packageId)
             .column(packageName)
             .column(repoName)
             .column(repoOwner)
+            .column(score)
             .column(summary)
             .from(searchView)
 
         return binds.reduce(preamble) { $0.where(haystack, contains, $1) }
             .where(isNotNull(packageName))
             .where(isNotNull(repoOwner))
             .where(isNotNull(repoName))
-            .orderBy(eq(lower(packageName), mergedTerms), .descending)
-            .orderBy(score, SQLDirection.descending)
-            .orderBy(packageName, SQLDirection.ascending)
+            .orderBy(sortOrder)
             .offset(offset)
             .limit(limit)
-            .select
+    }
+
+    static func keywordMatchQueryBuilder(on database: Database,
+                                         terms: [String]) -> SQLSelectBuilder {
+        guard let db = database as? SQLDatabase else {
+            fatalError("Database must be an SQLDatabase ('as? SQLDatabase' must succeed)")
+        }
+        let mergedTerms = SQLBind(terms.joined(separator: " ").lowercased())
+        // FIXME: bookend with `%`
```

Comment:
This initial implementation is doing a simple `"keyword" = 'test'` filtering with `LIMIT 1` when searching for `test`.

We'll want to also match `unit-testing`, changing that to `"keyword" like '%test%'` with some limit on the results, say `LIMIT 3`. When we do that we'll also need to `DISTINCT` the results (because the `UNNEST` will generate lots of duplicates).

More importantly, we'll need to define a sort order of partial keyword matches. I'm thinking something like

1. exact match
2. partial matches `%test%` ordered by number of occurrences, alphabetically

This will require quite a few additional changes, so I've left it out of this already big PR.